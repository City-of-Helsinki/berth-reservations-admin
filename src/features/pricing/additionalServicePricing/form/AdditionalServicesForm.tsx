import React from 'react';
import * as Yup from 'yup';
import { TextInput } from 'hds-react';
import { Field, Form, Formik } from 'formik';
import { useTranslation } from 'react-i18next';
import { TFunction } from 'i18next';

import Grid from '../../../../common/grid/Grid';
import Select from '../../../../common/select/Select';
import FormTypeTitle from '../../editModal/FormTypeTitle';
import {
  PeriodType,
  AdditionalProductTaxEnum,
  ProductServiceType,
  PriceUnits,
} from '../../../../@types/__generated__/globalTypes';
import {
  getPeriodTKey,
  getProductTax,
  getProductServiceTKey,
  getPriceUnits,
} from '../../../../common/utils/translations';
import Button from '../../../../common/button/Button';
import styles from './additionalServicesForm.module.scss';

const serviceOptions = Object.values(ProductServiceType);
const taxOptions = Object.values(AdditionalProductTaxEnum);
const priceUnitOptions = Object.values(PriceUnits);
const periodOptions = Object.values(PeriodType);

export interface AdditionalServiceValues {
  service: ProductServiceType | null;
  priceValue: number | string;
  priceUnit: PriceUnits | null;
  taxPercentage: AdditionalProductTaxEnum | null;
  period?: PeriodType | null;
}

export interface AdditionalServicesFormProps {
  initialValues?: AdditionalServiceValues | null;
  onSubmit(values: AdditionalServiceValues): void;
  onCancel(): void;
}

export const getAdditionalServicesValidationSchema = (t: TFunction) => {
  return Yup.object().shape({
    service: Yup.string().oneOf(serviceOptions).required(t('forms.common.errors.required')),
    priceValue: Yup.number()
      .positive()
      .typeError(t('forms.common.errors.numberType'))
      .required(t('forms.common.errors.required')),
    priceUnit: Yup.string().oneOf(priceUnitOptions).required(t('forms.common.errors.required')),
    taxPercentage: Yup.string()
      .oneOf(taxOptions)
      .typeError(t('forms.common.errors.numberType'))
      .required(t('forms.common.errors.required')),
    period: Yup.string().oneOf(periodOptions).required(t('forms.common.errors.required')),
  });
};

const OPTIONAL_SERVICES = [
  ProductServiceType.SUMMER_STORAGE_FOR_DOCKING_EQUIPMENT,
  ProductServiceType.SUMMER_STORAGE_FOR_TRAILERS,
  ProductServiceType.PARKING_PERMIT,
  ProductServiceType.DINGHY_PLACE,
];

const AdditionalServicesForm = ({ initialValues, onSubmit, onCancel }: AdditionalServicesFormProps) => {
  const { t, i18n } = useTranslation();
  const defaultValues: AdditionalServiceValues = {
    service: null,
    priceValue: '',
    priceUnit: null,
    taxPercentage: null,
    period: null,
  };

  return (
    <Formik
      initialValues={initialValues ?? defaultValues}
      onSubmit={onSubmit}
      validationSchema={getAdditionalServicesValidationSchema(t)}
    >
      {({ isSubmitting, errors, isValid, dirty }) => {
        const isSubmitDisabled = !dirty || isSubmitting || !isValid;

        return (
          <Form className={styles.form}>
            <div className={styles.row}>
              <FormTypeTitle label={t('common.terminology.dataEntry')} value={t('pricing.additionalServices.title')} />
            </div>
            <hr />
            <Grid colsCount={1} className={styles.row}>
              <Field
                required={true}
                as={Select}
                id="service"
                name="service"
                label={t('pricing.additionalServices.service')}
                disabled={initialValues?.service}
                options={serviceOptions
                  .filter((option) => OPTIONAL_SERVICES.includes(option))
                  .map((option) => ({
                    value: option,
                    label: t(getProductServiceTKey(option)),
                  }))}
              />
            </Grid>
            <Grid colsCount={3} className={styles.row}>
              <Field
                required={true}
                as={TextInput}
                id="priceValue"
                name="priceValue"
                label={t('pricing.additionalServices.price')}
                invalid={!!errors.priceValue}
                helperText={errors.priceValue}
              />
              <Field
                required={true}
                as={Select}
                id="priceUnit"
                name="priceUnit"
                label={t('pricing.additionalServices.unit')}
                options={priceUnitOptions.map((option) => ({
                  value: option,
                  label: getPriceUnits(option),
                }))}
              />
              <Field
                required={true}
                as={Select}
                id="taxPercentage"
                name="taxPercentage"
                label={t('pricing.additionalServices.tax')}
                options={taxOptions.map((option) => ({
                  value: option,
                  label: getProductTax(option, i18n.language),
                }))}
              />
            </Grid>
            <Grid colsCount={3} className={styles.row}>
              <Field
                required={true}
                as={Select}
                id="period"
                name="period"
                label={t('pricing.additionalServices.period')}
                options={periodOptions.map((option) => ({
                  value: option,
                  label: t(getPeriodTKey(option)),
                }))}
              />
            </Grid>
            <div className={styles.buttonRow}>
              <Button variant="secondary" id="cancel" onClick={onCancel}>
                {t('common.cancel')}
              </Button>
              <Button type="submit" disabled={isSubmitDisabled}>
                {t('common.save')}
              </Button>
            </div>
          </Form>
        );
      }}
    </Formik>
  );
};

export default AdditionalServicesForm;
