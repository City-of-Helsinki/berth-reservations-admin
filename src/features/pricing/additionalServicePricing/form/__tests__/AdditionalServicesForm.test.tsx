import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import AdditionalServicesForm, { AdditionalServicesFormProps } from '../AdditionalServicesForm';
import {
  ProductServiceType,
  PeriodType,
  AdditionalProductTaxEnum,
  PriceUnits,
} from '../../../../../@types/__generated__/globalTypes';

describe('AdditionalServicesForm', () => {
  const initialProps: AdditionalServicesFormProps = {
    onClose: jest.fn(),
    onSubmit: jest.fn(),
    initialValues: {
      service: ProductServiceType.LIGHTING,
      priceValue: 25,
      priceUnit: PriceUnits.AMOUNT,
      taxPercentage: AdditionalProductTaxEnum.TAX_24_00,
      period: PeriodType.SEASON,
    },
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  const getWrapper = (props: Partial<AdditionalServicesFormProps> = {}) =>
    mount(<AdditionalServicesForm {...initialProps} {...props} />);

  it('renders normally', () => {
    const wrapper = getWrapper();
    expect(wrapper.render()).toMatchSnapshot();
  });

  describe('Service field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('Select#service');
      expect(input.prop('value')).toEqual('LIGHTING');
    });
  });

  describe('Price Value field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('input#priceValue');
      expect(input.prop('value')).toEqual(25);
    });
  });

  describe('Price Unit field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('Select#priceUnit');
      expect(input.prop('value')).toEqual('AMOUNT');
    });
  });

  describe('Tax Percentage field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('Select#taxPercentage');
      expect(input.prop('value')).toEqual('TAX_24_00');
    });
  });

  describe('Period field', () => {
    it('Should show provided initialValue', () => {
      const input = getWrapper().find('Select#period');
      expect(input.prop('value')).toEqual('SEASON');
    });
  });

  describe('Modal buttons', () => {
    it('Should call the provided closeModal prop when cancel button is clicked', () => {
      const cancelBtn = getWrapper().find('Button').at(0);
      cancelBtn.simulate('click');

      expect(initialProps.onClose).toHaveBeenCalledTimes(1);
    });
    it('Should call the provided onSubmit prop with modal values when the form is submitted', async () => {
      const initialValues: AdditionalServicesFormProps['initialValues'] = {
        priceValue: 50,
        service: ProductServiceType.LIGHTING,
        priceUnit: PriceUnits.PERCENTAGE,
        taxPercentage: AdditionalProductTaxEnum.TAX_10_00,
        period: PeriodType.SEASON,
      };
      const submitBtn = getWrapper({ initialValues });

      await act(async () => {
        submitBtn.simulate('submit');
      });

      expect(initialProps.onSubmit).toHaveBeenNthCalledWith(1, initialValues, expect.anything());
    });
  });
});
