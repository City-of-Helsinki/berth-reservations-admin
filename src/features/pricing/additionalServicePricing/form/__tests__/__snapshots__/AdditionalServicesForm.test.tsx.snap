// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalServicesForm renders normally 1`] = `
<form
  action="#"
  class="form"
>
  <div
    class="heading"
  >
    <h4
      class="text brand h4"
    >
      Hinnaston muokkaus
    </h4>
  </div>
  <div
    class="row"
  >
    <div>
      <div
        class="formTypeTitleLabel"
      >
        Tietue
      </div>
      <div
        class="formTypeTitleValue"
      >
        Lisäpalvelut
      </div>
    </div>
  </div>
  <hr />
  <div
    class="grid cols1 row"
  >
    <div
      class="Select-module_root__Ka5uO Select-module_disabled__3MKDP"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="service-toggle-button"
        id="service-label"
      >
        Palvelu
        <span
          class="RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs"
        >
           *
        </span>
      </label>
      <div
        class="Select-module_wrapper__1WQXs"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="service-label service-toggle-button"
          aria-owns="service-menu"
          class="Select-module_button__1aIsm Select-module_placeholder__21c0j"
          disabled=""
          id="service-toggle-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Select-module_angleIcon__2-9AD"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h24v24H0z"
              />
              <path
                d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                fill="currentColor"
              />
            </g>
          </svg>
        </button>
        <ul
          aria-labelledby="service-label"
          aria-required="true"
          class="Select-module_menu__1H2aU Select-module_overflow__EZdsW"
          id="service-menu"
          role="listbox"
          style="max-height: 208px;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
  <div
    class="grid cols3 row"
  >
    <div
      class="TextInput-module_root__2CMNr text-input_hds-text-input__2LODq"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="priceValue"
      >
        Hinta
        <span
          class="RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs"
        >
           *
        </span>
      </label>
      <div
        class="TextInput-module_inputWrapper__3Rvel text-input_hds-text-input__input-wrapper__1OqYG"
      >
        <input
          class="TextInput-module_input__1BlHi text-input_hds-text-input__input__GJm5C"
          id="priceValue"
          name="priceValue"
          required=""
          type="text"
          value="25"
        />
      </div>
    </div>
    <div
      class="Select-module_root__Ka5uO"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="priceUnit-toggle-button"
        id="priceUnit-label"
      >
        Yksikkö
        <span
          class="RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs"
        >
           *
        </span>
      </label>
      <div
        class="Select-module_wrapper__1WQXs"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="priceUnit-label priceUnit-toggle-button"
          aria-owns="priceUnit-menu"
          class="Select-module_button__1aIsm"
          id="priceUnit-toggle-button"
          type="button"
        >
          <span
            class="Select-module_buttonLabel__1fqu5"
          >
            €
          </span>
          <svg
            aria-hidden="true"
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Select-module_angleIcon__2-9AD"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h24v24H0z"
              />
              <path
                d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                fill="currentColor"
              />
            </g>
          </svg>
        </button>
        <ul
          aria-labelledby="priceUnit-label"
          aria-required="true"
          class="Select-module_menu__1H2aU"
          id="priceUnit-menu"
          role="listbox"
          style="max-height: 260px;"
          tabindex="-1"
        />
      </div>
    </div>
    <div
      class="Select-module_root__Ka5uO"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="taxPercentage-toggle-button"
        id="taxPercentage-label"
      >
        ALV
        <span
          class="RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs"
        >
           *
        </span>
      </label>
      <div
        class="Select-module_wrapper__1WQXs"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="taxPercentage-label taxPercentage-toggle-button"
          aria-owns="taxPercentage-menu"
          class="Select-module_button__1aIsm"
          id="taxPercentage-toggle-button"
          type="button"
        >
          <span
            class="Select-module_buttonLabel__1fqu5"
          >
            24 %
          </span>
          <svg
            aria-hidden="true"
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Select-module_angleIcon__2-9AD"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h24v24H0z"
              />
              <path
                d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                fill="currentColor"
              />
            </g>
          </svg>
        </button>
        <ul
          aria-labelledby="taxPercentage-label"
          aria-required="true"
          class="Select-module_menu__1H2aU"
          id="taxPercentage-menu"
          role="listbox"
          style="max-height: 260px;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
  <div
    class="grid cols3 row"
  >
    <div
      class="Select-module_root__Ka5uO Select-module_disabled__3MKDP"
    >
      <label
        class="FieldLabel-module_label__1zrXK "
        for="period-toggle-button"
        id="period-label"
      >
        Aika
        <span
          class="RequiredIndicator-module_indicator__Pt7GX text-input_hds-text-input__required__2K-Bs"
        >
           *
        </span>
      </label>
      <div
        class="Select-module_wrapper__1WQXs"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="period-label period-toggle-button"
          aria-owns="period-menu"
          class="Select-module_button__1aIsm"
          disabled=""
          id="period-toggle-button"
          type="button"
        >
          <span
            class="Select-module_buttonLabel__1fqu5"
          >
            Kausi
          </span>
          <svg
            aria-hidden="true"
            class="Icon-module_icon__1Jtzj icon_hds-icon__1YqNC Icon-module_s__2WGWe icon_hds-icon--size-s__2Lkik Select-module_angleIcon__2-9AD"
            role="img"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="M0 0h24v24H0z"
              />
              <path
                d="M12 13.5l5-5 1.5 1.5-6.5 6.5L5.5 10 7 8.5z"
                fill="currentColor"
              />
            </g>
          </svg>
        </button>
        <ul
          aria-labelledby="period-label"
          aria-required="true"
          class="Select-module_menu__1H2aU"
          id="period-menu"
          role="listbox"
          style="max-height: 260px;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
  <div
    class="buttonRow"
  >
    <button
      class="Button-module_button__1msFE button_hds-button__2A0je Button-module_secondary__1nABp button_hds-button--secondary__1NOWS Button-module_theme-coat__2_3U7 button_hds-button--theme-coat__12cO0"
      id="cancel"
      type="button"
    >
      <span
        class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
      >
        Peruuta
      </span>
    </button>
    <button
      aria-disabled="true"
      class="Button-module_button__1msFE button_hds-button__2A0je Button-module_primary__2LfKB button_hds-button--primary__2NVvO Button-module_theme-coat__2_3U7 button_hds-button--theme-coat__12cO0"
      disabled=""
      type="submit"
    >
      <span
        class="Button-module_label__a4np1 button_hds-button__label__2EQa-"
      >
        Tallenna
      </span>
    </button>
  </div>
</form>
`;
